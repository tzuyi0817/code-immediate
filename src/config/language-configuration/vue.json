{
	"comments": {
		"blockComment": [
			"<!--",
			"-->"
		]
	},
	"brackets": [
		[
			"<!--",
			"-->"
		],
		[
			"<",
			">"
		],
		[
			"{",
			"}"
		],
		[
			"(",
			")"
		]
	],
	"autoClosingPairs": [
		{
			"open": "{",
			"close": "}"
		},
		{
			"open": "[",
			"close": "]"
		},
		{
			"open": "(",
			"close": ")"
		},
		{
			"open": "'",
			"close": "'"
		},
		{
			"open": "\"",
			"close": "\""
		},
		{
			"open": "<!--",
			"close": "-->",
			"notIn": [
				"comment",
				"string"
			]
		},
		{
			"open": "`",
			"close": "`",
			"notIn": [
				"string",
				"comment"
			]
		},
		{
			"open": "/**",
			"close": " */",
			"notIn": [
				"string"
			]
		}
	],
	"autoCloseBefore": ";:.,=}])><`'\" \n\t",
	"surroundingPairs": [
		{
			"open": "'",
			"close": "'"
		},
		{
			"open": "\"",
			"close": "\""
		},
		{
			"open": "{",
			"close": "}"
		},
		{
			"open": "[",
			"close": "]"
		},
		{
			"open": "(",
			"close": ")"
		},
		{
			"open": "<",
			"close": ">"
		},
		[
			"`",
			"`"
		]
	],
	"colorizedBracketPairs": [],
	"folding": {
		"markers": {
			"start": "^\\s*<!--\\s*#region\\b.*-->",
			"end": "^\\s*<!--\\s*#endregion\\b.*-->"
		}
	},
	"wordPattern": "(-?\\d*\\.\\d\\w*)|([^\\`\\~\\!\\@\\$\\^\\&\\*\\(\\)\\=\\+\\[\\{\\]\\}\\\\\\|\\;\\:\\'\\\"\\,\\.\\<\\>\\/\\s]+)",
	"indentationRules": {
		"increaseIndentPattern": "<(?!\\?|(?:area|base|br|col|frame|hr|html|img|input|keygen|link|menuitem|meta|param|source|track|wbr|script|style)\\b|[^>]*\\/>)([-_\\.A-Za-z0-9]+)(?=\\s|>)\\b[^>]*>(?!\\s*\\()(?!.*<\\/\\1>)|<!--(?!.*-->)|\\{[^}\"']*$",
		"decreaseIndentPattern": "^\\s*(<\\/(?!html)[-_\\.A-Za-z0-9]+\\b[^>]*>|-->|\\})"
	}
}