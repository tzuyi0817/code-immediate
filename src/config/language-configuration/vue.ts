import { languages } from 'monaco-editor';

export const vueConfiguration: languages.LanguageConfiguration = {
  comments: {
    blockComment: ['<!--', '-->'],
  },
  brackets: [
    ['<!--', '-->'],
    ['<', '>'],
    ['{', '}'],
    ['(', ')'],
  ],
  autoClosingPairs: [
    {
      open: '{',
      close: '}',
    },
    {
      open: '[',
      close: ']',
    },
    {
      open: '(',
      close: ')',
    },
    {
      open: "'",
      close: "'",
    },
    {
      open: '"',
      close: '"',
    },
    {
      open: '<!--',
      close: '-->',
      notIn: ['comment', 'string'],
    },
    {
      open: '`',
      close: '`',
      notIn: ['string', 'comment'],
    },
    {
      open: '/**',
      close: ' */',
      notIn: ['string'],
    },
  ],
  autoCloseBefore: ';:.,=}])><`\'" \n\t',
  surroundingPairs: [
    {
      open: "'",
      close: "'",
    },
    {
      open: '"',
      close: '"',
    },
    {
      open: '{',
      close: '}',
    },
    {
      open: '[',
      close: ']',
    },
    {
      open: '(',
      close: ')',
    },
    {
      open: '<',
      close: '>',
    },
    {
      open: '`',
      close: '`',
    },
  ],
  colorizedBracketPairs: [],
  folding: {
    markers: {
      start: /^\s*<!--\s*#region\b.*-->/,
      end: /^\s*<!--\s*#endregion\b.*-->/,
    },
  },
  wordPattern: /(-?\d*\.\d\w*)|([^\`\@\~\!\%\^\&\*\(\)\-\=\+\[\{\]\}\\\|\;\:\'\"\,\.\<\>/\?\s]+)/,
  onEnterRules: [
    {
      beforeText:
        /<(?!(?:area|base|br|col|embed|hr|img|input|keygen|link|menuitem|meta|param|source|track|wbr|script|style))([_:\w][_:\w-.\d]*)(?:(?:[^'"/>]|"[^"]*"|'[^']*')*?(?!\/)>)[^<]*$/i,
      afterText: /^<\/([_:\w][_:\w-.\d]*)\s*>/i,
      action: {
        indentAction: languages.IndentAction.IndentOutdent,
      },
    },
    {
      beforeText:
        /<(?!(?:area|base|br|col|embed|hr|img|input|keygen|link|menuitem|meta|param|source|track|wbr|script|style))([_:\w][_:\w-.\d]*)(?:(?:[^'"/>]|"[^"]*"|'[^']*')*?(?!\/)>)[^<]*$/i,
      action: {
        indentAction: languages.IndentAction.Indent,
      },
    },
  ],
  indentationRules: {
    increaseIndentPattern:
      /<(?!\?|area\b|base\b|br\b|col\b|frame\b|hr\b|html\b|img\b|input\b|keygen\b|link\b|menuitem\b|meta\b|param\b|source\b|track\b|wbr\b|script\b|style\b|[^>]*\/>)([-_.A-Za-z0-9]+)(?=\s|>)[^>]*>(?!\s*\()(?!.*<\/\1>)|<!--(?!.*-->)|\{[^}"']*$/,
    decreaseIndentPattern: /^\s*(<\/(?!html)[-_\.A-Za-z0-9]+\b[^>]*>|-->|\})/,
  },
};
