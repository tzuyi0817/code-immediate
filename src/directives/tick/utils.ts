import { RIPPLE_COUNT } from './constants';
import type { ITickDirectiveOptions } from './types';

export function createContainer({
  borderTopLeftRadius,
  borderTopRightRadius,
  borderBottomLeftRadius,
  borderBottomRightRadius,
}: CSSStyleDeclaration) {
  const tickContainer = document.createElement('div');

  tickContainer.style.top = '0';
  tickContainer.style.left = '0';
  tickContainer.style.width = '100%';
  tickContainer.style.height = '100%';
  tickContainer.style.position = 'absolute';

  tickContainer.style.borderRadius = `${borderTopLeftRadius} ${borderTopRightRadius} ${borderBottomRightRadius} ${borderBottomLeftRadius}`;
  tickContainer.style.overflow = 'hidden';
  tickContainer.style.pointerEvents = 'none';

  // Compatible with IE Apple
  tickContainer.style.webkitMaskImage = '-webkit-radial-gradient(white, black)';

  return tickContainer;
}

export function createTickElement(x: number, y: number, size: number, options: ITickDirectiveOptions) {
  const tickElement = document.createElement('div');

  tickElement.style.position = 'absolute';
  tickElement.style.width = `${size}px`;
  tickElement.style.height = `${size}px`;
  tickElement.style.top = `${y}px`;
  tickElement.style.left = `${x}px`;
  tickElement.style.background = options.color;
  tickElement.style.borderRadius = '50%';
  tickElement.style.opacity = `${options.initialOpacity}`;
  tickElement.style.transform = 'translate(-50%,-50%) scale(0)';
  tickElement.style.transition = `transform ${options.duration / 1000}s ${options.easing}, opacity ${
    options.duration / 1000
  }s ${options.easing}`;

  return tickElement;
}

export function magnitude(x1: number, y1: number, x2: number, y2: number): number {
  const deltaX = x1 - x2;
  const deltaY = y1 - y2;

  return Math.hypot(deltaX, deltaY);
}

export function getDistanceToFurthestCorner(x: number, y: number, { width, height }: DOMRect) {
  // Get the distance from the click target to the block scope boundary
  const topLeft = magnitude(x, y, 0, 0);
  const topRight = magnitude(x, y, width, 0);
  const bottomLeft = magnitude(x, y, 0, height);
  const bottomRight = magnitude(x, y, width, height);

  return Math.max(topLeft, topRight, bottomLeft, bottomRight);
}

export function getRelativePointer({ x, y }: PointerEvent, { top, left }: DOMRect) {
  return {
    x: x - left,
    y: y - top,
  };
}

function setTickCount(el: HTMLElement, count: number) {
  el.dataset[RIPPLE_COUNT] = count.toString();
}

export function getTickCount(el: HTMLElement): number {
  return Number.parseInt(el.dataset[RIPPLE_COUNT] ?? '0', 10);
}

export function incrementTickCount(el: HTMLElement): void {
  const count = getTickCount(el);

  setTickCount(el, count + 1);
}

export function decrementTickCount(el: HTMLElement): void {
  const count = getTickCount(el);

  setTickCount(el, count - 1);
}

export function deleteTickCount(el: HTMLElement): void {
  delete el.dataset[RIPPLE_COUNT];
}
