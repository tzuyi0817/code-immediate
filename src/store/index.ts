import { useCodeContentStore } from '@/store/modules/code-content';
import { useFlagStore } from '@/store/modules/flag';
import { useUserStore } from '@/store/modules/user';

export { useCodeContentStore, useFlagStore, useUserStore };
