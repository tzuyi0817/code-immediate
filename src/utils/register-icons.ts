import { library } from '@fortawesome/fontawesome-svg-core';
import { faCentos, faGithub } from '@fortawesome/free-brands-svg-icons';
import { faEye } from '@fortawesome/free-regular-svg-icons';
import {
  faAngleDown,
  faAngleLeft,
  faAngleRight,
  faAnglesLeft,
  faAnglesRight,
  faArrowRightFromBracket,
  faBarsStaggered,
  faCaretDown,
  faCheck,
  faCloudArrowUp,
  faEllipsis,
  faFileCirclePlus,
  faGear,
  faMagnifyingGlass,
  faPenFancy,
  faShare,
  faSheetPlastic,
  faSpinner,
  faTrash,
  faXmark,
} from '@fortawesome/free-solid-svg-icons';

export function registerIcons() {
  library.add(
    faPenFancy,
    faAngleDown,
    faSpinner,
    faCheck,
    faXmark,
    faGear,
    faMagnifyingGlass,
    faBarsStaggered,
    faCloudArrowUp,
    faSheetPlastic,
    faFileCirclePlus,
    faAngleLeft,
    faAnglesLeft,
    faAngleRight,
    faAnglesRight,
    faTrash,
    faEye,
    faGithub,
    faCentos,
    faArrowRightFromBracket,
    faShare,
    faCaretDown,
    faEllipsis,
  );
}
