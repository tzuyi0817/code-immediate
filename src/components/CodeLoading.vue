<script lang="ts" setup>
import { ref, computed, watch } from 'vue';
import { storeToRefs } from 'pinia';
import { useFlagStore } from '@/store';
import { sleep } from '@/utils/common';

const { isLoading, loadingType } = storeToRefs(useFlagStore());
const isShowTick = ref(false);
const isError = computed(() => loadingType.value.endsWith('error'));
const statusIcon = computed(() => `fa-${isError.value ? 'xmark' : 'check'}`);

async function flashTick() {
  isShowTick.value = true;
  await sleep(1000);
  isShowTick.value = false;
}

watch(isLoading, isOpen => !isOpen && flashTick());
</script>

<template>
  <div class="code_loading">
    <p
      v-if="isLoading || isShowTick"
      :class="['text-xs', { 'text-red-500': isError }]"
    >
      {{ loadingType }}
    </p>

    <font-awesome-icon
      v-if="isLoading"
      icon="fa-solid fa-spinner"
      title="fa-spinner"
      class="animate-spin code_loading_icon"
    />
    <font-awesome-icon
      v-if="isShowTick"
      :icon="`fa-solid ${statusIcon}`"
      :title="statusIcon"
      :class="['code_loading_icon', { 'text-red-500': isError }]"
    />
  </div>
</template>

<style lang="postcss" scoped>
.code_loading {
  @apply flex text-yellow-400 gap-2 font-mono;
}
</style>
