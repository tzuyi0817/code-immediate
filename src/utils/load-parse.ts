import loadjs from 'loadjs';
import { useFlagStore } from '@/store';
import { HTML_LANGUAGE_MAP, CSS_LANGUAGE_MAP, JS_LANGUAGE_MAP } from '@/config/language';
import type { CodeBase } from '@/types/code-content';

export const loadedParseMap = new Map([
  ['html', true],
  ['css', true],
  ['javascript', true],
]);

export function loadParse(language: string): Promise<void> {
  return new Promise((resolve, reject) => {
    if (loadedParseMap.get(language)) return resolve();
    const resource = `parses/${language}.js`;

    loadjs(resource, { returnPromise: true })
      .then(() => {
        loadedParseMap.set(language, true);
        resolve();
      })
      .catch((error: Error) => reject(error));
  });
}

export async function loadParseSource(language: string, languageMap: Record<string, string>) {
  const { setLoading } = useFlagStore();
  const source = languageMap[language];

  setLoading({ isOpen: true, type: 'Loading parse source' });

  if (source) {
    await loadParse(source).catch(error => {
      setLoading({ isOpen: false, type: 'Loading parse source error' });
      throw error;
    });
  }
  setLoading({ isOpen: false, type: 'Loading parse source finished' });
}

export function loadParseSources({ HTML, CSS, JS }: CodeBase) {
  return Promise.all([
    loadParseSource(HTML.language, HTML_LANGUAGE_MAP),
    loadParseSource(CSS.language, CSS_LANGUAGE_MAP),
    loadParseSource(JS.language, JS_LANGUAGE_MAP),
  ]);
}
