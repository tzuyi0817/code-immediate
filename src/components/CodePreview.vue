<script setup lang="ts">
import { ref, watch, inject, onMounted, type Ref } from 'vue';
import { storeToRefs } from 'pinia';
import { useCodeContentStore, useFlagStore } from '@/store';
import { createHtml } from '@/utils/create-html';
import { loadParseSources } from '@/utils/load-parse';
import { compile } from '@/utils/compile';
import { compileSfc } from '@/utils/compile-sfc';

const srcdoc = ref('');
const iframeRef: Ref<HTMLIFrameElement> | undefined = inject('iframeRef');
const { codeContent, isSFC } = storeToRefs(useCodeContentStore());
const { isFormatter, isStartDrag } = storeToRefs(useFlagStore());
const sandboxAttribute = [
  'allow-forms',
  'allow-modals',
  'allow-pointer-lock',
  'allow-popups',
  'allow-popups-to-escape-sandbox',
  'allow-same-origin',
  'allow-scripts',
  'allow-top-navigation-by-user-activation',
].join(' ');

async function runCode() {
  const { setLoading } = useFlagStore();
  const {
    codeContent: { HTML, CSS, JS, VUE },
    codeTemplate,
  } = useCodeContentStore();
  const compileFun = isSFC.value ? compileSfc : compile;

  setLoading({ isOpen: true, type: 'Process code' });

  try {
    const compileResult = await compileFun({
      html: {
        language: HTML.language,
        content: HTML.content,
      },
      css: {
        language: CSS.language,
        content: CSS.content,
        resources: CSS.resources,
      },
      js: {
        language: JS.language,
        content: JS.content,
        resources: JS.resources,
      },
      vue: {
        language: VUE.language,
        content: VUE.content,
      },
      codeTemplate,
    });

    srcdoc.value = createHtml({
      ...compileResult,
      cssResources: CSS.resources,
      jsResources: JS.resources,
    });
    setLoading({ isOpen: false, type: 'Process code finished' });
  } catch (error) {
    const isErrorConstructor = error instanceof Error;
    const message = isErrorConstructor ? error.message : String(error);

    iframeRef?.value?.contentWindow?.postMessage?.(
      {
        type: 'throwError',
        value: message,
      },
      '*',
    );
    setLoading({ isOpen: false, type: 'Process code error' });
    if (import.meta.env.MODE === 'test') return;
    throw new Error(message, { cause: isErrorConstructor ? error : void 0 });
  }
}

function initLoadParseSource() {
  const { HTML, CSS, JS } = codeContent.value;

  loadParseSources({ HTML, CSS, JS }).then(runCode);
}

watch(
  codeContent,
  () => {
    if (isFormatter.value) return;
    runCode();
  },
  { deep: true },
);

onMounted(initLoadParseSource);
</script>

<template>
  <div class="code_preview">
    <iframe
      ref="iframeRef"
      title="code preview"
      :sandbox="sandboxAttribute"
      :class="['h-full w-full', { 'pointer-events-none': isStartDrag }]"
      :srcdoc="srcdoc"
      frameborder="0"
    ></iframe>
  </div>
</template>
