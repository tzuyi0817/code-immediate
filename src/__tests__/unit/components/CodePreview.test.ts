import { ref, nextTick } from 'vue';
import { screen, fireEvent, waitFor } from '@testing-library/vue';
import CodePreview from '@/components/CodePreview.vue';
import CodeLoading from '@/components/CodeLoading.vue';
import { registerIcons } from '@/utils/register-icons';
import { useCodeContentStore } from '@/store';
import { loadedParseMap } from '@/utils/load-parse';
import { renderComponent } from '@/__tests__/unit/render';

describe('CodePreview component', () => {
  const codePreviewText = 'code preview';

  registerIcons();

  it('renders the correct content', async () => {
    renderComponent(CodePreview, { provide: { iframe: ref(null) } });
    expect(screen.getByTitle(codePreviewText)).toBeInTheDocument();
  });

  it('preview the correct content', async () => {
    const codeContentStore = useCodeContentStore();

    renderComponent(CodeLoading);
    renderComponent(CodePreview, { provide: { iframe: ref(null) } });
    await nextTick();
    codeContentStore.setCodeContent({ type: 'HTML', code: '<h1>Hello World!<h1>' });
    await fireEvent.load(screen.getByTitle(codePreviewText));
    await waitFor(() => {
      expect(screen.getByText('Process code finished')).toBeInTheDocument();
    });
    expect(screen.getByTitle(codePreviewText)).toMatchSnapshot();
  });

  it('catch preview compile error', async () => {
    const codeContentStore = useCodeContentStore();

    loadedParseMap.set('markdown', true);
    codeContentStore.setCodeLanguage({ type: 'HTML', language: 'Markdown' });
    renderComponent(CodeLoading);
    renderComponent(CodePreview, { provide: { iframe: ref(null) } });
    expect(await screen.findByText('Process code error')).toBeInTheDocument();
    loadedParseMap.delete('markdown');
  });
});
