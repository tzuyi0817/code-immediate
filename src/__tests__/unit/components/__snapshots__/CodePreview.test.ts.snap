// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CodePreview component > preview the correct content 1`] = `
<iframe
  class="h-full w-full"
  frameborder="0"
  sandbox="allow-forms allow-modals allow-pointer-lock allow-popups allow-popups-to-escape-sandbox allow-same-origin allow-scripts allow-top-navigation-by-user-activation"
  srcdoc="<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>code Demo</title>
    
    <script
      type="text/javascript"
      src="message/index.js"
    ></script>
    
    <script type="module">
      window.__modules__ = {};
      window.process = { env: { NODE_ENV: 'development' } };
    </script>
    
    
    <style>
      body {
        font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans',
          'Helvetica Neue', sans-serif;
      }
    </style>
    <style type="text/css"></style>
  </head>
  <body>
    <h1>Hello World!<h1>
    
  </body>
</html>
"
  title="code preview"
/>
`;
